guide_drafts/trait_derives.md

```diff
@@ -0,0 +1,927 @@
+
+```rust
+// File: src/models.rs
+
+use diesel::types::{Integer, Text};
+
+// We're defining the posts and users tables here
+// to illustrate the schema in the following example
+table! {
+    posts (id) {
+        id -> Int4,
+        user_id -> Int4,
+        title -> Varchar,
+        body -> Text,
+    }
+}
+
+table! {
+    users (id) {
+        id -> Int4,
+        first_name -> Varchar,
+        last_name -> Varchar,
+        email -> Varchar,
+    }
+}
+
+#[derive(Debug, Queryable)]
+pub struct User {
+    pub id: i32,
+    pub first_name: String,
+    pub last_name: String,
+    pub email: String,
+}
+
+// When not using the `table_name` annotation,
+// we must tell `QueryableByName` the type of
+// every field.
+#[derive(Debug, QueryableByName)]
+pub struct UserEmail {
+    #[sql_type="Integer"]
+    pub id: i32,
+    #[sql_type="Text"]
+    pub email: String,
+}
+
+// `full_name` is not a column on our users table,
+// but we're planning on returning a column that
+// concatenates the first and last name together.
+// `QueryableByName` will use the users table for
+// all other column types.
+#[derive(Debug, QueryableByName)]
+#[table_name="users"]
+pub struct UserName {
+    pub first_name: String,
+    pub last_name: String,
+    #[sql_type="Text"]
+    pub full_name: String,
+}
+
+// The user_name field maps to another
+// `QueryableByName` struct and all other fields
+// reference the posts table.
+#[derive(Debug, QueryableByName)]
+#[table_name="posts"]
+pub struct PostsWithUserName {
+    #[diesel(embed)]
+    pub user_name: UserName,
+    pub title: String,
+    pub body: String,
+}
+
+```
+
+```rust
+// File: src/main.rs
+
+use diesel::prelude::*;
+use diesel::sql_query;
+
+fn main() {
+    diesel::insert_into(users::table)
+        .values(
+          (
+            users::first_name.eq("Gordon"),
+            users::last_name.eq("Freeman"),
+            users::email.eq("gordon.freeman@blackmesa.co")
+          )
+        )
+        .execute(&db_connection)
+        .expect("Error inserting row into database");
+
+    let first_user = users
+        .first::<User>(&db_connection)
+        .expect("Error querying first user");
+
+    diesel::insert_into(posts::table)
+        .values(
+          (
+            posts::user_id.eq(first_user.id),
+            posts::title.eq("Thoughts on Tomorrow's Experiment"),
+            posts::body.eq("What could possibly go wrong?")
+          )
+        )
+        .execute(&db_connection)
+        .expect("Error inserting row into database");
+
+    let users_emails = sql_query("SELECT users.id, users.email FROM users ORDER BY id")
+      .load::<UserEmail>(&connection);
+    println!("{:?}", users_emails); 
+    // > User { id: 1, email: "gordon.freeman@blackmesa.co" }
+    
+    let joined = sql_query("SELECT users.first_name, users.last_name, CONCAT(users.first_name, users.last_name) as full_name, posts.body, posts.title FROM users INNER JOIN posts ON users.id = posts.user_id")
```

Comment:
@sgrif I'm not 100% convinced this is a great example, as I don't want to mislead readers about joins. However, it does illustrate that `QueryableByName`:
- doesn't have to be restricted to a single table
- May reference columns not available in schema
- Supports nested `QueryableByName` structs
- You can mix the annotations (table_name / sql_type)

Is there a common use case that might illustrate this better?

Reply:
I think this is good. If you are looking to make it _simpler, maybe make a

```rust
struct User {
    #[diesel(embed)]
    name: UserName,
    email: String
}
```

(with your UserName struct), and do this query (maybe even incl. the formatting):

```sql
SELECT first_name,
       last_name,
       CONCAT(users.first_name, users.last_name) AS full_name,
       email
  FROM users
```

It doesn't show the table name fun, but is much shorter.

You should probably add a comment in both cases that this is just an example and that you could also do this with the query builder.